Further clarification on new grains docs from #37028

terminalmage opened the pull request.

terminalmage committed:
Further clarification on new grains docs from #37028

rallytime commented on doc/topics/grains/index.rst:
```diff
@@ -133,20 +133,30 @@ Writing Grains
+be designed to support many platforms, operating systems or applications. Also,
+consider whether or not the information is likely to change. Grains are for the
+most part static data that should rarely (if ever) change. Information that is
+more dynamic should be retrived by an execution module. Also, keep in mind that
```
`retrieved` instead of `retrived`

rallytime commented on doc/topics/grains/index.rst:
```diff
@@ -133,20 +133,30 @@ Writing Grains
+
+    ...
+    ...
+    {{ salt['module.function_name']('argument_1, 'argument_2') }}
```
Missing a closing quote on `argument_1`

rallytime commented on doc/topics/grains/index.rst:
```diff
@@ -133,20 +133,30 @@ Writing Grains
+in the grains package or the custom grains directory. The functions in the
+modules of the grains must return a Python :ref:`dict <python2:typesmapping>`,
+where the keys in the :ref:`dict <python2:typesmapping>` are the names of the
+grains and the values are the values.
```
Maybe change this last part to say something a little more clear like `and the dictionary values are the grain data values` or something? I had to read this last part of the sentence twice because I stumble on what values were the values and what they were for. That might just be me though. *shrugs*

terminalmage replied:
I just pushed a commit which makes the corrections you suggested, and also restructures things a bit. In the process of going back to make the edits, I noticed some redundancies, so I moved some things around and reworded a few things as well. The diff may look odd now, it might be better to read the file in its [updated state](https://raw.githubusercontent.com/terminalmage/salt/a0502a7b907fe999d3ba299d5e45baddc2b0986a/doc/topics/grains/index.rst).

rallytime reviewed (commented):
Just a couple of minor things, and one suggestion. Feel free to take it or leave my suggestion at the end.

terminalmage committed:
Restructure grain writing docs

Some information recently added is a little redundant. This commit edits
it and restructures the information to make it flow better.

rallytime reviewed (approved):
Awesome! Thanks Erik!

terminalmage committed:
Add one more paragraph

damon-atkins commented:
When to Use a Custom Grain       since this is a separate section now, I suggest its moved a bit higher before the detail of where to place files in _grain.

I suggest you place the comment about "Minimal" back.  Otherwise large amounts data will be included because once you start adding the code its easy to include a lots of info.   For example zpool grain has been added recently which includes the size of pool.   It should be enough to for the names of the zpools to be supplied or even zpools are in use.  I am sure their would be other existing grains which include to much information.  Their needs to be guidance on the amount of data to included.    This 2nd reason for the update was to give guidance on the amount of data or what to included.

An example of a new grains would be    list of volume manages in use  or a list of file system types in use.

terminalmage commented:
It's not a separate section, it's a subsection underneath ``Writing Grains``. Given that, I think it's fine where it is. The first few paragraphs in the ``Writing Grains`` section explain how to write them and where to put them, and the two subsections underneath offer more helpful information for those who want to write custom grains.

Your suggestion to put "minimal data" in grains is, frankly, just an arbitrary assertion. If you want to observe that as a rule of thumb, then great, but I'm not going to include it in the docs and make it a general guideline without getting more feedback from the development team.

cachedout commented:
I have to agree with @terminalmage here. While targeting is certainly the common case, It's not clear to me why a grain should _only_ be used for targeting when it could be used for whatever a user desires. (I'm aware of a number of users with perfectly valid use cases that have nothing to do with targeting and others who produce lots of information from grains.) It's also not clear to me why a grain should be be designed to "support" multiple platforms when plenty of grains are platform-specific.

terminalmage commented:
@cachedout I do agree that a grain might not necessarily be written to support multiple platforms, but there are cases where it should, and we should probably do a better job in the docs of making the distinction of when that should be a consideration.

damon-atkins commented:
_gain more up to what the user wants, some guidance does help and "core grain" needs to be a bit more restrictive.

The reason I started the doco improvement came out of  #36941 "state.highstate command triggering grain refresh"

>It's also not clear to me why a grain should be be designed to "support" multiple platforms when plenty of grains are platform-specific.

It does use the word "should"

I would have thought the idea of targeting with grains to be non-platform specific.
e.g. a gain which indicated valid virtual modules active, so you don't get an error back from something which does not support module abc.

Gains going into core should be about "How can I make this platform/os independent, if appropriate" so its usefulness is much greater across more platforms.    Maybe gains for core should have more guidance in Contribution doco.
